pkg/db/db_session/default.go

```diff
@@ -149,21 +149,26 @@ func waitForNotification(ctx context.Context, l *pq.Listener, callback func(id s
 			if n != nil {
 				logger.V(4).Infof("Received event from channel [%s] : %s", n.Channel, n.Extra)
 				callback(n.Extra)
+			} else {
+				// nil notification means the connection was closed
+				logger.Infof("recreate the listener due to the connection loss")
+				l.Close()
+				// recreate the listener
+				l = newListener(ctx, dbConfig, channel)
 			}
 		case <-time.After(10 * time.Second):
 			logger.V(10).Infof("Received no events on channel during interval. Pinging source")
-			go func() {
-				// TODO: Need to handle the error, especially in cases of network failure.
-				err := l.Ping()
-				if err != nil {
-					logger.Error(err.Error())
-				}
-			}()
+			if err := l.Ping(); err != nil {
+				logger.Infof("recreate the listener due to %s", err.Error())
+				l.Close()
+				// recreate the listener
+				l = newListener(ctx, dbConfig, channel)
```

Comment:
will this be a problem? it seems we are changing a shared variable without lock. What happened when another is using the listener, and it is changed here?

Reply:
if reaching here, it means the listener has problem. it cannot be used anymore.

Reply:
what happens when a caller calls the listener at the same time, will it panic?

Reply:
the listener is initialized when the maestro server is starting. https://github.com/openshift-online/maestro/blob/e28d0430f167086ea3ee590575187a1e2a8e6cce/cmd/maestro/server/controllers.go#L61
so there is no new caller to call the listener. we leverage LISTEN/NOTIFY mechanism  to sends a notification event together with an optional “payload” string to listener that has previously executed LISTEN channel.